hnix-store-core/src/System/Nix/Internal/Path.hs

```diff
@@ -0,0 +1,150 @@
+import qualified System.Nix.Hash
+
+-- | Parse store location
+parseStore :: BSL.ByteString -> T.Text
+parseStore = T.pack . takeDirectory . BSC.unpack . BSL.toStrict
+
+-- | Parse path from string
+--
+-- in form <storeDir>/<hash>-<pathName>
+-- into (Just (Path <hash> (PathName <pathName>)))
+-- or Nothing on error
+--
+-- XXX: should check for @PathHashAlgo length
+parsePath :: BSL.ByteString -> Maybe Path
+parsePath p = case name of
+    Nothing -> Nothing
+    Just n -> Just $ Path digest n
+  where
+  base = T.pack . takeBaseName . BSC.unpack . BSL.toStrict $ p
+  parts = T.breakOn "-" base
+  digest = Digest . BSC.pack . T.unpack . fst $ parts
+  name = pathName . T.drop 1 . snd $ parts
+
+
+-- experimental
+-- Directory of the store
+type StoreDir = Text
+type Stored a = (StoreDir, a)
+
+-- wrap StoreDir and Path into tuple
+makeStored :: StoreDir -> Path -> Stored Path
+makeStored sl p = (sl, p)
+
+type PathType = Text
+-- "text:<r1>:<r2>:...<rN>"
+-- "source"
+-- "output:<id>"
+--    <id> is the name of the output (usually, "out").
+
+-- store settings
+data Settings = Settings {
+  storeDir :: StoreDir -- settings.nixStore
+  } deriving (Eq, Show)
+
+-- build a store path in the following form:
+-- <storeDir>/<hash>-<pathName>
+storedToText :: Stored Path -> Text
+storedToText (storeLoc, (Path digest pName)) = T.concat
+  [ storeLoc
+  , "/"
+  , printAsBase32 @PathHashAlgo digest
+  , "-"
+  , pathNameContents pName
+  ]
+
+makeStorePath :: (HasDigest a) => PathType -> PathName -> Digest a -> Settings -> Text
+makeStorePath typ pName digest settings = T.concat
+  [ storeDir settings
+  , "/"
+  , printAsBase32 @PathHashAlgo $ pathHash typ pName digest (storeDir settings)
+  , "-"
+  , pathNameContents pName
+  ]
+
+makeStorePath' :: (HasDigest a) => PathType -> PathName -> Digest a -> StoreDir -> Path
+makeStorePath' typ pName digest storeLoc = snd $ makeStoredPath typ pName digest storeLoc
+
+-- | build Stored Path from the type of the path, path name and a digest stored at StoreDir
+-- As StoreDir is part of the path hashing process we need to take it into account
+-- when building Path(s)
+makeStoredPath :: (HasDigest a) => PathType -> PathName -> Digest a -> StoreDir -> Stored Path
+makeStoredPath typ pName digest storeLoc = makeStored storeLoc $ Path (pathHash typ pName digest storeLoc) pName
+
+-- build <h> string which is a truncated base32 formatted SHA256 hash of <s>
+pathHash :: (HasDigest a) => PathType -> PathName -> Digest a -> StoreDir -> Digest PathHashAlgo
+pathHash typ pName digest storeLoc = System.Nix.Hash.hash . BSC.pack . T.unpack $
+  makePathDigestString typ pName digest storeLoc
+
+-- build <s> string which is hashed and used in makeStorePath
+-- <s> = "<pathType>:<hash_algo>:<base16_hash>:<storeDir>:<pathName>"
+-- (exposed for testing purposes only)
+makePathDigestString ::  (HasDigest a) => PathType -> PathName -> Digest a -> StoreDir -> Text
```

Comment:
If this function is only used in `makeStorePath`, can we move it into a `where` clause in `makeStorePath`? That could help to reduce the largish number of functions with similar-sounding names at the top level.